avocado/core/task/statemachine.py

```diff
@@ -74,20 +75,69 @@ async def bootstrap(self):
 
     async def triage(self):
         """Reads from triaging, moves into either: ready or finished."""
+        def fail_on_triage(runtime_task):
+            self._state_machine.finished.append(runtime_task)
+            runtime_task.status = 'FAILED ON TRIAGE'
+
         try:
             async with self._state_machine.lock:
                 runtime_task = self._state_machine.triaging.pop(0)
         except IndexError:
             return
 
-        requirements_ok = await self._spawner.check_task_requirements(runtime_task)
-        if requirements_ok:
+        # check for requirements a task may have
+        requirements_ok = (
+                await self._spawner.check_task_requirements(runtime_task))
+        if not requirements_ok:
             async with self._state_machine.lock:
-                self._state_machine.ready.append(runtime_task)
-        else:
+                fail_on_triage(runtime_task)
+            return
+
+        # handle task dependencies
+        if runtime_task.task.dependencies:
+            # check of all the dependency tasks finished
             async with self._state_machine.lock:
-                self._state_machine.finished.append(runtime_task)
-                runtime_task.status = 'FAILED ON TRIAGE'
+                finished_tasks = [rt_task.task for rt_task
+                                  in self._state_machine.finished]
+            if not runtime_task.task.dependencies.issubset(finished_tasks):
+                async with self._state_machine.lock:
+                    self._state_machine.triaging.append(runtime_task)
+                    runtime_task.status = 'WAITING DEPENDENCIES'
+                await asyncio.sleep(0.1)
+                return
+
+            # dependencies finished, let's check if they finished
+            # successfully, so we can move on with the parent task
+            for task in runtime_task.task.dependencies:
+                # check if this dependency `task` failed on triage
+                # this is needed because this kind of task fail does not
+                # have information in the status repo
+                for finished_rt_task in self._state_machine.finished:
+                    if finished_rt_task.task is task:
+                        if finished_rt_task.status == 'FAILED ON TRIAGE':
```

Comment:
nitpick: To improve readability I would suggest adding a `AND` here.

Reply:
I did it in the first implementation, but the line exceeds the 80 column limit, so I thought it would be more readable, breaking it into 2 `if` statements. Anyway, I don't have a problem merging them and breaking the line.

Reply:
I'm merging the lines and sending the self-test fix in a force push.